Create reusable `SelectionMapFragment`

seadowg opened the pull request:
Work towards #4947 

There's been a bunch of discussions about whether we use an Activity or a Fragment as the reusable component here, and we decided a Fragment made the most sense. There's still a bunch of tidying up and solidifying to do but this felt like a good point to get a review before I go any further.

Next steps:

1. Convert `FormMapActivity` to Kotlin
1. Distribute tests driving out form map between a (rewritten) `FormMapActivityTest` and new tests for `SelectionMapFragment` and `SelectionSummarySheet`
1. Remove `FormMapViewModel` (making sure the above covers all of its logic)
1. Convert `FormMapActivity` to Kotlin
1. Reevaluate need for `FormMapTest` (reworking the tests might make it redundant or at least allow us to delete some of the tests)
1. Investigate moving shared map behaviour from `SelectionMapFragment` into `MapFragment` implementations (maintaining map zoom on recreate for instance)
1. Investigate extracting `maps` module to house `MapFragment`, `MapPoint` and `MapFragmentFactory` and creating modules for each implementation (classic **D**epenednecy **I**nversion **P**rinciple setup)

#### What has been done to verify that this works as intended?

Managed to keep existing tests working (with some tweaks to make the boundaries less sporadic) and added a couple of new ones where it there was no coverage to safely move code around. Verified manually as well.

#### Why is this the best possible solution? Were any other approaches considered?

Discussed on Slack.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

Again, this shouldn't result in any changes to behaviour. I think we can just merge this without QA (and look at the whole feature again when we're finished).

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

seadowg committed:
Remove remaining uses of ANSWER_KEY

seadowg committed:
Extract summary sheet view away from form related logic

seadowg committed:
Stop using internal field for tests

seadowg committed:
Use MappableSelectItem as much as possible internally without changing ViewModel

seadowg committed:
Move selection and items state over to new ViewModel

seadowg committed:
Move multi click safe components to shared module

seadowg committed:
Move all FormMapViewModel interactions into overload methods

seadowg committed:
Pull most logic up into SelectionMapActivity

seadowg committed:
Move Intent setup to abstract method

seadowg committed:
Switch to using ViewModel to manage SelectionMapActivity state

seadowg committed:
Inline FormMapViewModel method

seadowg committed:
Move to Fragment implementation of reusabl selection map

seadowg committed:
Deprecate FormMapViewModel

seadowg committed:
Remove unused view model swap out code

seadowg committed:
Fix summary sheet width

seadowg committed:
Add missing permission check for opening form map

seadowg commented on collect_app/src/main/java/org/odk/collect/android/application/Collect.java:
```diff
@@ -282,7 +283,7 @@ public ReferenceLayerSettingsNavigator providesReferenceLayerSettingsNavigator()
                         @Override
                         public MapFragmentFactory providesMapFragmentFactory() {
-                            return applicationComponent.mapProvider();
+                            return applicationComponent.mapFragmentFactory();
```
I think we should really be using the `MapFragmentFactory` interface wherever this in injected. `MapProvider` is just the implementation of that interface.

seadowg committed:
Fix dependency resolution for release builds

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/utilities/IconUtils.java:
```diff
@@ -52,17 +52,18 @@ public static Bitmap getBitmap(Context context, int drawableId) {
     }
 
-    public static Drawable getSubmissionSummaryStatusIcon(Context context, String instanceStatus) {
+    public static int getSubmissionSummaryStatusIcon(String instanceStatus) {
```
I think `IconUtils` is not a good place for this method so maybe we could move it?

seadowg replied:
Yeah, good point. I'll work out somewhere else for it to live.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/SelectionMapFragment.kt:
```diff
@@ -0,0 +1,416 @@
+import org.odk.collect.permissions.PermissionsProvider
+import javax.inject.Inject
+
+class SelectionMapFragment : Fragment() {
```
Now as it is a fragment we should clear binding `onDestroyView()` https://developer.android.com/topic/libraries/view-binding#fragments

seadowg replied:
Huh interesting. My thinking was that there would be no need for this as recreation would simply replace the reference. There's a suggestion in [this SO](https://stackoverflow.com/questions/65295104/android-view-binding-clear-binding-in-fragment-lifecycle) that one potential reason to `null` out the reference is to prevent leaks if the `Fragment` hangs around after the view is destroyed. However, it doesn't seem like that's a case anyone has a tangible example of (and [this linked article](https://www.techyourchance.com/android-fragment-lifecycle/) even refers to the practice as "premature optimization". But of course, the official docs do state that: 

> Fragments outlive their views. Make sure you clean up any references to the binding class instance in the fragment's [onDestroyView()](https://developer.android.com/reference/kotlin/androidx/fragment/app/Fragment#ondestroyview) method.

It honestly feels like View Binding wasn't though out for Fragments properly (like everything else), and they've had to add that in due to memory problems people have run into. It's not even in their [code samples](https://github.com/android/architecture-components-samples/blob/main/GithubBrowserSample/app/src/main/java/com/android/example/github/ui/user/UserFragment.kt) for instance.

After getting grumpy about all this and looking through Android's issue tracker I found [this discussion](https://issuetracker.google.com/issues/152333892) that links [this article](https://issuetracker.google.com/issues/152333892) that demonstrates using`<binding class>.bind(view)` in `onViewCreated` to avoid create binding fields. I'm going to file an issue around that being added to the View Binding docs (https://developer.android.com/topic/libraries/view-binding) and go with that to avoid needing an `onDestroyView`.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/SelectionMapFragment.kt:
```diff
@@ -0,0 +1,416 @@
+        const val RESULT_CREATE_NEW_ITEM = "create_new_item"
+
+        private val MAP_CENTER_KEY = "map_center"
+        private val MAP_ZOOM_KEY = "map_zoom"
```
Those two might be consts.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/SelectionMapFragment.kt:
```diff
@@ -0,0 +1,416 @@
+package org.odk.collect.geo
```
Now in this file we have a few classes  what makes it big are you going to split it?

seadowg replied:
Yeah it feels messy. How about we discuss the package structure as part of the next PR, so it's as closed to "finished" as possible? That way we hopefully avoid any churn on things moving around.

grzesiek2010 replied:
Ok

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/SelectionMapFragment.kt:
```diff
@@ -0,0 +1,416 @@
+                if (featureId == selectionMapViewModel.getSelectedItemId()) item.largeIcon else item.smallIcon
+            )
+
+            itemsByFeatureId.put(featureId, item)
```
Could be `itemsByFeatureId[featureId] = item`

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/SelectionMapFragment.kt:
```diff
@@ -0,0 +1,416 @@
+        }
+    }
+
+    private fun onActionClick(view: View) {
```
view is unused.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/SelectionMapFragment.kt:
```diff
@@ -0,0 +1,416 @@
+    private var mapTitle = MutableLiveData<String>()
+    private var mappableItems = MutableLiveData<List<MappableSelectItem>>(emptyList())
+    private var itemCount = MutableLiveData(0)
+    private var selectedItemId = -1
```
Can we make it a const? (that -1) or use null to represent this state I remember you saying recently that you prefer null over -1 like here

seadowg replied:
Oh thanks yes! I'd meant to clean this up but it escaped me.

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/SelectionMapFragment.kt:
```diff
@@ -0,0 +1,416 @@
+
+        updateFeatures(items)
+
+        if (!viewportInitialized && !points.isEmpty()) {
```
You can use `points.isNotEmpty()` instead of `!points.isEmpty()`

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/SelectionMapFragment.kt:
```diff
@@ -0,0 +1,416 @@
+    }
+
+    private fun update(items: List<MappableSelectItem>) {
+        if (map == null) {
```
`map` is lateinint so it shouldn't be null?

grzesiek2010 commented on geo/src/main/java/org/odk/collect/geo/SelectionMapFragment.kt:
```diff
@@ -0,0 +1,416 @@
+
+    private fun onClick() {
+        if (summarySheetBehavior.getState() == BottomSheetBehavior.STATE_EXPANDED) {
+            summarySheetBehavior.setState(BottomSheetBehavior.STATE_HIDDEN)
```
You can call just state. Please fix other `setSate` too.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FillBlankFormActivity.kt:
```diff
@@ -218,11 +197,15 @@ class FillBlankFormActivity :
-                    selectFormFromMap.launch(
-                        Input(currentProjectProvider.getCurrentProject().uuid, id)
-                    )
+                    val intent =
```
This could be just:
```
Intent(this@FillBlankFormActivity, FormMapActivity::class.java).also {
                        it.putExtra(FormMapActivity.EXTRA_FORM_ID, id)
                        startActivity(it)
                    }
```

instead of:

```                   
 val intent =
                        Intent(this@FillBlankFormActivity, FormMapActivity::class.java).also {
                            it.putExtra(FormMapActivity.EXTRA_FORM_ID, id)
                        }

                    startActivity(intent)

```

seadowg replied:
I went with inlining `Intent` instead. I think it's good to make sure `startActivity` is front and center as it's the code here's main feature.

grzesiek2010 replied:
I would use my syntax but ok it's a matter of taste.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/activities/FillBlankFormActivity.kt:
```diff
@@ -218,11 +197,15 @@ class FillBlankFormActivity :
             this,
-            object : PermissionListener {
+            object :
+                PermissionListener {
```
We don't have any check for this but to me it's weird to have that in two lines I would rather have: 
`object : PermissionListener {`

seadowg committed:
Move method to only class it's used in

seadowg committed:
Remove binding field to avoid need to clean it up in onDestroyView

seadowg committed:
Improve converted Kotlin code

seadowg committed:
Fix formatting

seadowg committed:
Use correct check for lateinit

seadowg committed:
Switch to optional

seadowg committed:
Update test that flaked out to use page helpers

I could have updated the whole file but that feels like a pretty big
shift in focus from the selection map right now.
